x-pack/plugins/apm/server/routes/services/get_service_container_metadata.ts

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { ElasticsearchClient } from '@kbn/core/server';
+import { rangeQuery } from '@kbn/observability-plugin/server';
+import {
+  CONTAINER,
+  CONTAINER_ID,
+  CONTAINER_IMAGE,
+  KUBERNETES,
+  KUBERNETES_CONTAINER_NAME,
+  KUBERNETES_NAMESPACE,
+  KUBERNETES_POD_NAME,
+  KUBERNETES_POD_UID,
+  KUBERNETES_REPLICASET_NAME,
+  KUBERNETES_DEPLOYMENT_NAME,
+} from '../../../common/elasticsearch_fieldnames';
+import { Kubernetes } from '../../../typings/es_schemas/raw/fields/kubernetes';
+import { Container } from '../../../typings/es_schemas/raw/fields/container';
+import { maybe } from '../../../common/utils/maybe';
+
+export interface ContainerMetadata {
+  kubernetes?: Kubernetes;
+  container?: Container;
+}
+
+interface ResponseHitSource {
+  [key: string]: {
+    pod: {
+      name: string;
+      uid: string;
+    };
+    image: {
+      name: string;
+    };
+    id: string;
+  };
+}
+interface ResponseAggregations {
+  [key: string]: {
+    buckets: Array<{
+      key: string;
+    }>;
+  };
+}
+
+export const getServiceContainerMetadata = async ({
+  esClient,
+  indexName,
+  containerIds,
+  start,
+  end,
+}: {
+  esClient: ElasticsearchClient;
+  indexName?: string;
+  containerIds: string[];
+  start: number;
+  end: number;
+}): Promise<ContainerMetadata> => {
+  if (!indexName) {
+    return {};
+  }
+
+  const should = [
+    { exists: { field: KUBERNETES } },
+    { exists: { field: CONTAINER_IMAGE } },
+    { exists: { field: KUBERNETES_CONTAINER_NAME } },
+    { exists: { field: KUBERNETES_NAMESPACE } },
+    { exists: { field: KUBERNETES_POD_NAME } },
+    { exists: { field: KUBERNETES_POD_UID } },
+    { exists: { field: KUBERNETES_REPLICASET_NAME } },
+    { exists: { field: KUBERNETES_DEPLOYMENT_NAME } },
+  ];
+
+  const response = await esClient.search<unknown, ResponseAggregations>({
+    index: [indexName],
+    _source: [KUBERNETES, CONTAINER],
+    query: {
+      bool: {
+        filter: [
+          {
+            terms: {
+              [CONTAINER_ID]: containerIds,
+            },
+          },
+          ...rangeQuery(start, end),
+        ],
+        should,
+      },
+    },
+    aggs: {
+      deployment: {
+        terms: {
+          field: KUBERNETES_DEPLOYMENT_NAME,
+          size: 10,
+        },
+      },
+      namespace: {
+        terms: {
+          field: KUBERNETES_NAMESPACE,
+          size: 10,
+        },
+      },
+      replicaset: {
+        terms: {
+          field: KUBERNETES_REPLICASET_NAME,
+          size: 10,
+        },
+      },
+    },
+  });
+
+  const sources = maybe(response.hits.hits[0])?._source as ResponseHitSource;
```

Comment:
Are we just getting the first source doc? What about the aggs? Are we not using those or am I missing something?

Reply:
Oh, I was blind. I see now that we are indeed using the aggs below. 
But why are we using the first source document?

Reply:
Used `top_metrics` aggs: ffede596f7267c9e1a8919d83049a53651eec75e

Reply:
top_metrics returns the values from the top n documents (which are not guaranteed to contain all unique values), I think you need the top n unique values from all documents? in which case a terms agg is what you need

Reply:
I believe the confusion started as I used one query to fetch 2 different things:

1. Kubernetes metadata for the service overview (_many containers_)
2. Kubernetes metadata for the service node (_specific container_)

**For case 1**: I need the n unique values from all documents. (`terms`)
**For case 2**: I need sample data as it's metadata and I _expect_ it should be the same in all documents (or that's not safe assumption?) (`hits.hits[0]`)

Cleaned up the code for **case 1**: https://github.com/elastic/kibana/pull/139612/commits/ff1296cc7dc98d334ca4ea9b64e4c237ff5da003 as it should refer to plural for example
Created a new query for **case 2**: https://github.com/elastic/kibana/pull/139612/commits/a4518c49dbdc3f30e32358de2e38256cdba3265a

Reply:
@kpatticha you're using a `should` clause, which (in this case) means the documents will be sorted by how well that match the should clause (and its queries). It might not be the same for all documents, but you'll get the best-matching document as the first hit, so that looks fine.